### ISSUE TITLE
Scuba on M1

### ISSUE THREAD

**Issue by christheyankee** [JonathonReinhart/scuba#196]
Scuba does not seem to work on the M1 macs right now.

Reproduction configuration
```
image: debian:11
```
```
$ scuba ls
[drwxr-xr-x 2 root root 4096 Dec  6 13:09 missing
touch: cannot touch 'missing/test': Permission denied](docker: Error response from daemon: failed to create shim task: OCI runtime create failed: runc create failed: unable to start container process: exec: "/.scuba/scubainit": stat /.scuba/scubainit: no such file or directory: unknown.)
```
Expected Behavior
Something similar to:
```
$ docker run -it --rm -v $(pwd):/test:ro -w /test -u $(id -u):$(id -g) debian:11 ls
Backend  Dockerfile  Frontend  README.md
```

**Comment by Kfeavel** [comment]
I use `scuba` with my M1 Pro Mac. The issue is that `scuba` does not work with Debian based images, but other, non-Debian images work fine. Try using `scuba` with an `alpine` image and it should work just fine.

**Comment by christheyankee** [comment]
Seems to fail with alpine as well

**Comment by Kfeavel** [comment]
Can you try with `ghcr.io/xyrisos/xyris-build/xyris-build:5.0.0`? I've been using that this week and it's worked fine for me.

**Comment by christheyankee** [comment]
fails with that as well

**Comment by Kfeavel** [comment]
Oh, interesting. How do you have Docker installed? I'm using Docker Desktop.

**Comment by JonathonReinhart** [comment]
In general, Suba is not verified to work on Mac.  I don't know how Docker works on Mac these days.

```
stat /.scuba/scubainit: no such file or directory
```

This makes me suspect that the bind-mounting is not working. That wouldn't surprise me, because that's where we had trouble with OSX in the past. IIRC, only certain host directories could be bind-mounted (presumably due the underlying mechanism used to share directories with the Linux VM) -- and when you specified a directory that wouldn't work, it failed silently, resulting in an empty directory. We bind-mount the `.scuba` dir, so that looks like a strong possibility of what's happening here. It might depend on how/where you have scuba installed on the host. Try from your home directory or something.

I have neither the bandwidth nor a platform on which to work on this. If anyone wants to contribute, please feel free to volunteer!

**Comment by christheyankee** [comment]
> Oh, interesting. How do you have Docker installed? I'm using Docker Desktop.

Yeah iirc i just ran smth like `brew install docker --cask`

**Comment by JonathonReinhart** [comment]
If you want to debug this further, I suggest trying `scuba --dry-run` ([docs](https://scuba.readthedocs.io/en/latest/cli.html)). This will do two things:

1. Generate the temporary files.
2. Dump out the scuba command line it would run.

You can copy / paste that and play around.

I suspect that if you run it with `ls /.scuba`, you'll see that `scubainit` is absent. Other things may or may not be there, depending on whhere they are on the host. You can play around and see what works.

If you figure it out, and can ideally point to documentation about what can(not) be bind-mounted on OSX, I would appreciate a patch to at least detect this problem, and point users at the docs where it is explained and a workaround is provided.

**Comment by christheyankee** [comment]
yeah the directory exists but its empty. ill look into whats going on a bit fs.

**Comment by christheyankee** [comment]
Okay so after some investigation its that on macs `/var` is a symlink to `private/var` after resolving that debian images are stuck on #109 but other images seem to work.

**Comment by JonathonReinhart** [comment]
> Okay so after some investigation its user error. It escaped me that on macs `/var` is a symlink to `/private/var`. After resolving that

@christheyankee Can you please elaborate? What exactly was the "user error"? How does the fact that `/var` is a symlink to `/private/var` cause this issue? How did you resolve it?

Chances are that other Mac users are likely to run into this same problem, so anything we can do now to illuminate the problem saves future hassle for multiple people.

**Comment by JonathonReinhart** [comment]
> The issue is that scuba does not work with Debian based images

@Kfeavel That is a very broad, and concerning statement. [`debian:10` is used](https://github.com/JonathonReinhart/scuba/blob/v2.10.0/tests/const.py#L10) for all of the main tests, so it is definitely not entirely accurate.

If you are aware of a particular failure mode, please open an issue. There may be some specific configuration or subset of images that do not work, but to say "Scuba does not work with Debian based images" without referring to a specific problem is potentially damaging.

**Comment by JonathonReinhart** [comment]
#109 applies only to images that were built with [`USER`](https://docs.docker.com/engine/reference/builder/#user) in their Dockerfile. If `docker inspect -f '{{ .Config.User }}' $image` doesn't output anything, then #109 doesn't apply.

I've moved this debian-images conversation to #197.

**Comment by christheyankee** [comment]
> > Okay so after some investigation its user error. It escaped me that on macs `/var` is a symlink to `/private/var`. After resolving that
> 
> @christheyankee Can you please elaborate? What exactly was the "user error"? How does the fact that `/var` is a symlink to `/private/var` cause this issue? How did you resolve it?
> 
> Chances are that other Mac users are likely to run into this same problem, so anything we can do now to illuminate the problem saves future hassle for multiple people.

The user error is that you have to allow `/private/var/folders` in Docker Desktop under `Preferences -> Resources -> File sharing`. The problem is that when you run with `--dry-run` you get output that appears to create resources in `/var/folder`. I then added that to the Docker Desktop resources as described above and the issue persisted. It was only after I realized it was a symlink and then added the real path that it started working.

**Comment by christheyankee** [comment]
> > The issue is that scuba does not work with Debian based images
> 
> @Kfeavel That is a very broad, and concerning statement. [`debian:10` is used](https://github.com/JonathonReinhart/scuba/blob/v2.10.0/tests/const.py#L10) for all of the main tests, so it is definitely not entirely accurate.
> 
> If you are aware of a particular failure mode, please open an issue. There may be some specific configuration or subset of images that do not work, but to say "Scuba does not work with Debian based images" without referring to a specific problem is potentially damaging.

It works fine if you delete the `staff` group.

**Comment by Kfeavel** [comment]
> @Kfeavel That is a very broad, and concerning statement. [debian:10 is used](https://github.com/JonathonReinhart/scuba/blob/v2.10.0/tests/const.py#L10) for all of the main tests, so it is definitely not entirely accurate.

Apologies, I should have clarified that this issue is specific to macOS. I have attempted to use Debian images with scuba on macOS for quite some time now and have never had success. I should have been more specific when making my statement.

**Comment by JonathonReinhart** [comment]
Would either of you care to submit a PR which updates the docs to include battle-tested instructions for getting Scuba working on a Mac?

Specifically:
- What configuration must be set in Docker Desktop? (https://github.com/JonathonReinhart/scuba/issues/196#issuecomment-1374254866)
- Are there limitations on the directories from which scuba can be run?

I've never owned a Mac and it has been many years since scuba attempted to actually "support" running on a Mac, so I'm both impressed and confused that it works. If you could also clarify this for me, that'd be great:

## History

I don't know exactly how Docker on Mac works today, but back in the 2016-2017 timeframe, I believe Docker ("Desktop"?) on OSX ran a `boot2docker` VM, and then set the `DOCKER_HOST` environment variable to pipe all local docker commands to that machine. It had some fairly rigid directory-sharing set up so things pretty much worked if you ran Scuba from a project in the right directory. Because of this, a lot of the `scubainit` stuff I added was unnecessary (the files ended up being owned on the host by your user anyway.) So I added code to basically say "if `DOCKER_HOST` is set, don't use `scubainit`: 4beadf2b3a29c8bf2bac1b321b88b4851bb197b0

https://github.com/JonathonReinhart/scuba/blob/4beadf2b3a29c8bf2bac1b321b88b4851bb197b0/src/scuba#L279-L292

But eventually that one Mac user didn't care anymore, and I had no way of testing it, so I dropped support for it in a refactor leading up to Scuba 2.0: a918d448d54ee3c8f93487eb5b9cd8a30f361870. Now if `DOCKER_HOST` was set, Scuba would throw its hands up in the air. That code is still there today: https://github.com/JonathonReinhart/scuba/blob/b086413404a8d002d9d25d18ac50dc8bf940a062/scuba/scuba.py#L90-L100

Given that you guys say it works, `DOCKER_HOST` must not be set, which means:
- Scuba thinks it is operating in "native" mode (see `__setup_native_run()`). At a minimum this is misleading, but at worst it introduces some inconsistencies vs running truly native on Linux.
- Presumably, the docker daemon is running locally, and connected to via a local UNIX socket? But Docker runs Linux executables, so there is presumably a Linux kernel running somewhere... How does folder sharing into this presumed VM work?

**Comment by christheyankee** [comment]
> How does folder sharing into this presumed VM work?

From a cursory scan of [docker.com](https://www.docker.com/blog/deep-dive-into-new-docker-desktop-filesharing-implementation/) it looks like a FUSE client and server over Hypervisor sockets. 

> Given that you guys say it works, DOCKER_HOST must not be set

That would line up with what I'm seeing.
```
$ echo $DOCKER_HOST

$
```

> Would either of you care to submit a PR which updates the docs to include battle-tested instructions for getting Scuba working on a Mac?

I'd be happy to do this but I think it might be better to hold off until we decide what to do about #197 since Docker is such a common choice for docker images.

**Comment by MatthewTingum** [comment]
You might find yourself at this issue if you "snap" installed docker. Error message in OP is similar if not identical. This can cause permissions issues (maybe I haven't really tracked it down). Try installing docker from your apt repo (`docker.io` or whatever).

**Comment by retu2libc** [comment]
That error is because when docker is installed via snap, it is sandboxed to the user's $HOME directory. See [here](https://stackoverflow.com/questions/65267251/docker-bind-mount-directory-in-tmp-not-working).
